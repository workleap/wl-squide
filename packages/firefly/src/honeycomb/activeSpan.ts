import type { Span } from "@opentelemetry/api";
import { isPlainObject } from "@squide/core";
import type { RootLogger } from "@workleap/logging";
import { v4 as uuidv4 } from "uuid";
import { createTraceContextId } from "./createTraceContextId.ts";

export type ActiveSpanId = string;

export interface ActiveSpan {
    id: ActiveSpanId;
    name: string;
    instance: Span;
}

// Using a stack because we want a Last In First Out implementation for this.
// https://github.com/open-telemetry/opentelemetry-js/issues/5084
// https://github.com/open-telemetry/opentelemetry-js/issues/3558#issuecomment-1760680244
class ActiveSpanStack {
    readonly #stack: ActiveSpan[] = [];

    push(span: ActiveSpan) {
        this.#stack.push(span);
    }

    pop(span: ActiveSpan) {
        const head = this.#stack.pop();

        if (!head) {
            throw new Error("[squide] Unexpected pop, the active Honeycomb span stack is empty.");
        }

        if (head.id !== span.id) {
            throw new Error(`[squide] The active Honeycomb span is not the expected span. Expected to pop span with name and id "${span.name} / ${span.id}" but found "${head.name} / ${head.id}". Did you forget to end an active span?`);
        }

        return head;
    }

    peek() {
        if (this.#stack.length === 0) {
            return undefined;
        }

        return this.#stack[this.#stack.length - 1];
    }
}

const GlobalActiveSpanStackVariableName = "__SQUIDE_HONEYCOMB_ACTIVE_SPAN_STACK__";

export function registerActiveSpanStack() {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    if (globalThis[GlobalActiveSpanStackVariableName]) {
        throw new Error(`[squide] An ActiveSpanStack instance has already been registered to globalThis.${GlobalActiveSpanStackVariableName}. Did you register the Honeycomb instrumentation twice?`);
    }

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    globalThis[GlobalActiveSpanStackVariableName] = new ActiveSpanStack();
}

function getActiveSpanStack() {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    return globalThis[GlobalActiveSpanStackVariableName] as ActiveSpanStack;
}

function getActiveSpan() {
    const stack = getActiveSpanStack();

    if (stack) {
        return stack.peek();
    }
}

export function setActiveSpan(name: string, span: Span) {
    const activeSpan: ActiveSpan = {
        id: uuidv4(),
        name: name,
        instance: span
    };

    const stack = getActiveSpanStack();

    if (stack) {
        stack.push(activeSpan);
    }

    return activeSpan;
}

export function popActiveSpan(span: ActiveSpan) {
    const stack = getActiveSpanStack();

    if (stack) {
        stack.pop(span);
    }
}

export function createOverrideFetchRequestSpanWithActiveSpanContext(logger: RootLogger) {
    return (span: Span, request: Request | RequestInit) => {
        const activeSpan = getActiveSpan();

        if (activeSpan) {
            const activeSpanContext = activeSpan.instance.spanContext();
            const requestSpanContext = span.spanContext();

            if (activeSpanContext) {
                logger
                    .withText("[squide] Found a Honeycomb active context to apply to the following fetch request:")
                    .withLineChange()
                    .withText("Request span context:")
                    .withObject(requestSpanContext)
                    .withLineChange()
                    .withText("Active span context:")
                    .withObject(activeSpanContext)
                    .withLineChange()
                    .withText("Request:")
                    .withObject(request)
                    .debug();

                span.setAttribute("trace.trace_id", activeSpanContext.traceId);
                span.setAttribute("trace.parent_id", activeSpanContext.spanId);

                const traceParent = createTraceContextId(activeSpanContext.traceId, requestSpanContext.spanId, requestSpanContext.traceFlags);

                if (request instanceof Request) {
                    request.headers.set("traceparent", traceParent);
                } else if (isPlainObject(request.headers)) {
                    request.headers["traceparent"] = traceParent;
                }

                // Indicates to not propagate the requests to the subsequent hooks.
                return true;
            }
        }
    };
}
