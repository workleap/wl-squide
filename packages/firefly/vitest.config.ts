import react from "@vitejs/plugin-react";
import { defineConfig } from "vitest/config";

export default defineConfig({
    test: {
        environment: "happy-dom",
        include: ["tests/**/*.test.{ts,tsx}"],
        exclude: ["node_modules", "dist"],
        setupFiles: ["./vitest-setup.ts"],
        testTransformMode: {
            web: [".ts", ".tsx"]
        },
        reporters: "verbose"
    },
    define: {
        __webpack_share_scopes__: {
            default: {}
        }
    },
    cacheDir: "./node_modules/.cache/vitest",
    plugins: [react()]
});
