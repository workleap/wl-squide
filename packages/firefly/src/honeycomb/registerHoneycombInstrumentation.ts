import type { Span } from "@opentelemetry/api";
import {
    type AddListenerOptions,
    type EventCallbackFunction,
    type EventName,
    LocalModuleDeferredRegistrationFailedEvent,
    LocalModuleDeferredRegistrationUpdateFailedEvent,
    LocalModuleRegistrationFailedEvent,
    LocalModulesDeferredRegistrationCompletedEvent,
    type LocalModulesDeferredRegistrationCompletedEventPayload,
    LocalModulesDeferredRegistrationStartedEvent,
    type LocalModulesDeferredRegistrationStartedEventPayload,
    LocalModulesDeferredRegistrationsUpdateCompletedEvent,
    type LocalModulesDeferredRegistrationsUpdateCompletedEventPayload,
    LocalModulesDeferredRegistrationsUpdateStartedEvent,
    type LocalModulesDeferredRegistrationsUpdateStartedEventPayload,
    LocalModulesRegistrationCompletedEvent,
    type LocalModulesRegistrationCompletedEventPayload,
    LocalModulesRegistrationStartedEvent,
    type LocalModulesRegistrationStartedEventPayload,
    type ModuleRegistrationError
} from "@squide/core";
import {
    DeferredRegistrationsUpdateCompletedEvent,
    DeferredRegistrationsUpdateStartedEvent,
    RemoteModuleDeferredRegistrationFailedEvent,
    RemoteModuleDeferredRegistrationUpdateFailedEvent,
    type RemoteModuleRegistrationError,
    RemoteModuleRegistrationFailedEvent,
    RemoteModulesDeferredRegistrationCompletedEvent,
    type RemoteModulesDeferredRegistrationCompletedEventPayload,
    RemoteModulesDeferredRegistrationStartedEvent,
    type RemoteModulesDeferredRegistrationStartedEventPayload,
    RemoteModulesDeferredRegistrationsUpdateCompletedEvent,
    type RemoteModulesDeferredRegistrationsUpdateCompletedEventPayload,
    RemoteModulesDeferredRegistrationsUpdateStartedEvent,
    type RemoteModulesDeferredRegistrationsUpdateStartedEventPayload,
    RemoteModulesRegistrationCompletedEvent,
    type RemoteModulesRegistrationCompletedEventPayload,
    RemoteModulesRegistrationStartedEvent,
    type RemoteModulesRegistrationStartedEventPayload
} from "@squide/module-federation";
import { ApplicationBoostrappedEvent, type AppRouterWaitState, ModulesReadyEvent, ModulesRegisteredEvent, MswReadyEvent, ProtectedDataReadyEvent, PublicDataReadyEvent } from "../AppRouterReducer.ts";
import type { FireflyRuntime } from "../FireflyRuntime.tsx";
import { ApplicationBootstrappingStartedEvent } from "../initializeFirefly.ts";
import { ProtectedDataFetchFailedEvent, ProtectedDataFetchStartedEvent } from "../useProtectedDataQueries.ts";
import { PublicDataFetchFailedEvent, PublicDataFetchStartedEvent } from "../usePublicDataQueries.ts";
import { type ActiveSpan, createOverrideFetchRequestSpanWithActiveSpanContext, registerActiveSpanStack } from "./activeSpan.ts";
import { getTracer } from "./tracer.ts";
import { endActiveSpan, startActiveChildSpan, startChildSpan, startSpan, traceError } from "./utils.ts";

// TIPS:
// To query those traces in Honeycomb, use the following query filter: "root.name = squide-bootstrapping".

interface AddProtectedListenerOptions extends AddListenerOptions {
    onError?: (error: unknown) => void;
}

function addProtectedListener(runtime: FireflyRuntime, eventName: EventName, callback: EventCallbackFunction, options?: AddProtectedListenerOptions) {
    const protectedCallback = (...args: unknown[]) => {
        try {
            callback(...args);
        } catch (error: unknown) {
            runtime.logger
                .withText(`[squide] An unmanaged error occurred while handling event "${eventName.toString()}" for Honeycomb instrumentation:`)
                .withError(error as Error)
                .error();
        }
    };

    runtime.eventBus.addListener(eventName, protectedCallback, options);
}

type DataFetchState = "none" | "fetching-data" | "public-data-ready" | "protected-data-ready" | "data-ready" | "data-fetch-failed";

export function reduceDataFetchEvents(
    runtime: FireflyRuntime,
    onDataFetchStarted: () => void,
    onDataReady: () => void,
    onPublicDataFetchStarted: () => void,
    onPublicDataReady: () => void,
    onProtectedDataFetchStarted: () => void,
    onProtectedDataReady: () => void,
    onDataFetchFailed: (queriesErrors: Error[]) => void,
    onUnmanagedError: (error: unknown) => void
) {
    let dataFetchState: DataFetchState = "none";

    addProtectedListener(runtime, PublicDataFetchStartedEvent, () => {
        if (dataFetchState === "none") {
            dataFetchState = "fetching-data";
            onDataFetchStarted();
        }

        onPublicDataFetchStarted();
    }, {
        once: true,
        onError: onUnmanagedError
    });

    addProtectedListener(runtime, PublicDataReadyEvent, payload => {
        onPublicDataReady();

        if (dataFetchState === "fetching-data") {
            if (payload && !(payload as AppRouterWaitState).waitForProtectedData) {
                dataFetchState = "data-ready";
                onDataReady();
            } else {
                dataFetchState = "public-data-ready";
            }
        } else if (dataFetchState === "protected-data-ready") {
            dataFetchState = "data-ready";
            onDataReady();
        }
    }, {
        once: true,
        onError: onUnmanagedError
    });

    addProtectedListener(runtime, ProtectedDataFetchStartedEvent, () => {
        if (dataFetchState === "none") {
            dataFetchState = "fetching-data";
            onDataFetchStarted();
        }

        onProtectedDataFetchStarted();
    }, {
        once: true,
        onError: onUnmanagedError
    });

    addProtectedListener(runtime, ProtectedDataReadyEvent, payload => {
        onProtectedDataReady();

        if (dataFetchState === "fetching-data") {
            if (payload && !(payload as AppRouterWaitState).waitForPublicData) {
                dataFetchState = "data-ready";
                onDataReady();
            } else {
                dataFetchState = "protected-data-ready";
            }
        } else if (dataFetchState === "public-data-ready") {
            dataFetchState = "data-ready";
            onDataReady();
        }
    }, {
        once: true,
        onError: onUnmanagedError
    });

    const handleDataFetchFailed = (payload: unknown) => {
        if (dataFetchState !== "data-fetch-failed") {
            dataFetchState = "data-fetch-failed";

            onDataFetchFailed(payload as Error[]);
        }
    };

    addProtectedListener(runtime, PublicDataFetchFailedEvent, handleDataFetchFailed, {
        once: true,
        onError: onUnmanagedError
    });

    addProtectedListener(runtime, ProtectedDataFetchFailedEvent, handleDataFetchFailed, {
        once: true,
        onError: onUnmanagedError
    });
}

function registerTrackingListeners(runtime: FireflyRuntime) {
    let bootstrappingSpan: Span;
    let bootstrappingSpanHasEnded: boolean = false;
    let localModuleRegistrationSpan: Span;
    let localModuleDeferredRegistrationSpan: Span;
    let remoteModuleRegistrationSpan: Span;
    let remoteModuleDeferredRegistrationSpan: Span;
    let dataFetchSpan: ActiveSpan;
    let deferredRegistrationsUpdateSpan: Span;
    let localModuleDeferredRegistrationsUpdateSpan: ActiveSpan;
    let remoteModuleDeferredRegistrationsUpdateSpan: ActiveSpan;

    const handleUnmanagedError = (error: unknown) => {
        if (bootstrappingSpan && !bootstrappingSpanHasEnded) {
            traceError(bootstrappingSpan, error as Error);

            bootstrappingSpan.end();
            bootstrappingSpanHasEnded = true;
        }

        if (localModuleRegistrationSpan) {
            localModuleRegistrationSpan.end();
        }

        if (localModuleDeferredRegistrationSpan) {
            localModuleDeferredRegistrationSpan.end();
        }

        if (remoteModuleRegistrationSpan) {
            remoteModuleRegistrationSpan.end();
        }

        if (remoteModuleDeferredRegistrationSpan) {
            remoteModuleDeferredRegistrationSpan.end();
        }

        if (dataFetchSpan) {
            dataFetchSpan.instance.end();
        }

        if (deferredRegistrationsUpdateSpan) {
            deferredRegistrationsUpdateSpan.end();
        }

        if (localModuleDeferredRegistrationsUpdateSpan) {
            localModuleDeferredRegistrationsUpdateSpan.instance.end();
        }

        if (remoteModuleDeferredRegistrationsUpdateSpan) {
            remoteModuleDeferredRegistrationsUpdateSpan.instance.end();
        }
    };

    addProtectedListener(runtime, ApplicationBootstrappingStartedEvent, () => {
        bootstrappingSpan = startSpan((options, context) => getTracer().startSpan("squide-bootstrapping", options, context));
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, ApplicationBoostrappedEvent, () => {
        if (bootstrappingSpan) {
            bootstrappingSpan.end();
            bootstrappingSpanHasEnded = true;
        }
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, MswReadyEvent, () => {
        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("msw-ready");
        }
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, LocalModulesRegistrationStartedEvent, (payload: unknown) => {
        const attributes = {
            "app.squide.module_count": (payload as LocalModulesRegistrationStartedEventPayload).moduleCount
        };

        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("local-module-registration-started", attributes);
        }

        localModuleRegistrationSpan = startChildSpan(bootstrappingSpan, (options, context) => {
            return getTracer().startSpan("local-module-registration", { ...options, attributes }, context);
        });
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, LocalModulesRegistrationCompletedEvent, (payload: unknown) => {
        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("local-module-registration-completed", {
                "app.squide.module_count": (payload as LocalModulesRegistrationCompletedEventPayload).moduleCount
            });
        }

        if (localModuleRegistrationSpan) {
            localModuleRegistrationSpan.end();
        }
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, LocalModuleRegistrationFailedEvent, (payload: unknown) => {
        const registrationError = payload as ModuleRegistrationError;

        if (localModuleRegistrationSpan) {
            traceError(localModuleRegistrationSpan, registrationError);
        }
    }, {
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, LocalModulesDeferredRegistrationStartedEvent, (payload: unknown) => {
        const attributes = {
            "app.squide.registration_count": (payload as LocalModulesDeferredRegistrationStartedEventPayload).registrationCount
        };

        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("local-module-deferred-registration-started", attributes);
        }

        localModuleDeferredRegistrationSpan = startChildSpan(bootstrappingSpan, (options, context) => {
            return getTracer().startSpan("local-module-deferred-registration", { ...options, attributes }, context);
        });
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, LocalModulesDeferredRegistrationCompletedEvent, (payload: unknown) => {
        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("local-module-deferred-registration-completed", {
                "app.squide.registration_count": (payload as LocalModulesDeferredRegistrationCompletedEventPayload).registrationCount
            });
        }

        if (localModuleDeferredRegistrationSpan) {
            localModuleDeferredRegistrationSpan.end();
        }
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, LocalModuleDeferredRegistrationFailedEvent, (payload: unknown) => {
        const registrationError = payload as ModuleRegistrationError;

        if (localModuleDeferredRegistrationSpan) {
            traceError(localModuleRegistrationSpan, registrationError);
        }
    }, {
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, RemoteModulesRegistrationStartedEvent, (payload: unknown) => {
        const attributes = {
            "app.squide.remote_count": (payload as RemoteModulesRegistrationStartedEventPayload).remoteCount
        };

        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("remote-module-registration-started", attributes);
        }

        remoteModuleRegistrationSpan = startChildSpan(bootstrappingSpan, (options, context) => {
            return getTracer().startSpan("remote-module-registration", { ...options, attributes }, context);
        });
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, RemoteModulesRegistrationCompletedEvent, (payload: unknown) => {
        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("remote-module-registration-completed", {
                "app.squide.remote_count": (payload as RemoteModulesRegistrationCompletedEventPayload).remoteCount
            });
        }

        if (remoteModuleRegistrationSpan) {
            remoteModuleRegistrationSpan.end();
        }
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, RemoteModuleRegistrationFailedEvent, (payload: unknown) => {
        const registrationError = payload as RemoteModuleRegistrationError;

        if (remoteModuleRegistrationSpan) {
            traceError(remoteModuleRegistrationSpan, registrationError);
        }
    }, {
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, RemoteModulesDeferredRegistrationStartedEvent, (payload: unknown) => {
        const attributes = {
            "app.squide.registration_count": (payload as RemoteModulesDeferredRegistrationStartedEventPayload).registrationCount
        };

        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("remote-module-deferred-registration-started", attributes);
        }

        remoteModuleDeferredRegistrationSpan = startChildSpan(bootstrappingSpan, (options, context) => {
            return getTracer().startSpan("remote-module-deferred-registration", { ...options, attributes }, context);
        });
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, RemoteModulesDeferredRegistrationCompletedEvent, (payload: unknown) => {
        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("remote-module-deferred-registration-completed", {
                "app.squide.registration_count": (payload as RemoteModulesDeferredRegistrationCompletedEventPayload).registrationCount
            });
        }

        if (remoteModuleDeferredRegistrationSpan) {
            remoteModuleDeferredRegistrationSpan.end();
        }
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, RemoteModuleDeferredRegistrationFailedEvent, (payload: unknown) => {
        const registrationError = payload as RemoteModuleRegistrationError;

        if (remoteModuleDeferredRegistrationSpan) {
            traceError(remoteModuleDeferredRegistrationSpan, registrationError);
        }
    }, {
        onError: handleUnmanagedError
    });

    const handleFetchDataStarted = () => {
        dataFetchSpan = startActiveChildSpan(bootstrappingSpan, (options, context) => {
            const name = "data-fetch";
            const span = getTracer().startSpan(name, options, context);

            return {
                name,
                span
            };
        });
    };

    const handleDataReady = () => {
        if (dataFetchSpan) {
            endActiveSpan(dataFetchSpan);
        }
    };

    const handlePublicDataFetchStarted = () => {
        if (dataFetchSpan) {
            dataFetchSpan.instance.addEvent("public-data-fetch-started");
        }
    };

    const handlePublicDataReady = () => {
        if (dataFetchSpan) {
            dataFetchSpan.instance.addEvent("public-data-ready");
        }
    };

    const handleProtectedDataFetchStarted = () => {
        if (dataFetchSpan) {
            dataFetchSpan.instance.addEvent("protected-data-fetch-started");
        }
    };

    const handleProtectedDataReady = () => {
        if (dataFetchSpan) {
            dataFetchSpan.instance.addEvent("protected-data-ready");
        }
    };

    const handleDataFetchFailed = (queriesErrors: Error[]) => {
        if (dataFetchSpan) {
            queriesErrors.forEach(x => {
                traceError(dataFetchSpan.instance, x);
            });

            endActiveSpan(dataFetchSpan);

            // Global data fetch errors are unmanaged, which mean the host application bootstrapping flow
            // will be aborted and a react-router error boundary will be rendered.
            if (bootstrappingSpan) {
                bootstrappingSpan.end();
                bootstrappingSpanHasEnded = true;
            }
        }
    };

    reduceDataFetchEvents(
        runtime,
        handleFetchDataStarted,
        handleDataReady,
        handlePublicDataFetchStarted,
        handlePublicDataReady,
        handleProtectedDataFetchStarted,
        handleProtectedDataReady,
        handleDataFetchFailed,
        handleUnmanagedError
    );

    addProtectedListener(runtime, ModulesRegisteredEvent, () => {
        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("modules-registered");
        }
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    addProtectedListener(runtime, ModulesReadyEvent, () => {
        if (bootstrappingSpan) {
            bootstrappingSpan.addEvent("modules-ready");
        }
    }, {
        once: true,
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, DeferredRegistrationsUpdateStartedEvent, () => {
        deferredRegistrationsUpdateSpan = startSpan((options, context) => getTracer().startSpan("squide-deferred-registrations-update", options, context));
    }, {
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, DeferredRegistrationsUpdateCompletedEvent, () => {
        if (deferredRegistrationsUpdateSpan) {
            deferredRegistrationsUpdateSpan.end();
        }
    }, {
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, LocalModulesDeferredRegistrationsUpdateStartedEvent, (payload: unknown) => {
        const attributes = {
            "app.squide.registration_count": (payload as LocalModulesDeferredRegistrationsUpdateStartedEventPayload).registrationCount
        };

        if (deferredRegistrationsUpdateSpan) {
            deferredRegistrationsUpdateSpan.addEvent("local-module-deferred-registrations-update-started", attributes);
        }

        localModuleDeferredRegistrationsUpdateSpan = startActiveChildSpan(deferredRegistrationsUpdateSpan, (options, context) => {
            const name = "local-module-deferred-registrations-update";

            const span = getTracer().startSpan(name, {
                attributes,
                ...options
            }, context);

            return {
                name,
                span
            };
        });
    }, {
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, LocalModulesDeferredRegistrationsUpdateCompletedEvent, (payload: unknown) => {
        if (deferredRegistrationsUpdateSpan) {
            deferredRegistrationsUpdateSpan.addEvent("local-module-deferred-registrations-update-completed", {
                "app.squide.registration_count": (payload as LocalModulesDeferredRegistrationsUpdateCompletedEventPayload).registrationCount
            });
        }

        if (localModuleDeferredRegistrationsUpdateSpan) {
            endActiveSpan(localModuleDeferredRegistrationsUpdateSpan);
        }
    }, {
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, LocalModuleDeferredRegistrationUpdateFailedEvent, (payload: unknown) => {
        const registrationError = payload as ModuleRegistrationError;

        if (localModuleDeferredRegistrationsUpdateSpan) {
            traceError(localModuleDeferredRegistrationsUpdateSpan.instance, registrationError);
        }
    }, {
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, RemoteModulesDeferredRegistrationsUpdateStartedEvent, (payload: unknown) => {
        const attributes = {
            "app.squide.registration_count": (payload as RemoteModulesDeferredRegistrationsUpdateStartedEventPayload).registrationCount
        };

        if (deferredRegistrationsUpdateSpan) {
            deferredRegistrationsUpdateSpan.addEvent("remote-module-deferred-registrations-update-started", attributes);
        }

        remoteModuleDeferredRegistrationsUpdateSpan = startActiveChildSpan(deferredRegistrationsUpdateSpan, (options, context) => {
            const name = "remote-module-deferred-registrations-update";

            const span = getTracer().startSpan(name, {
                attributes,
                ...options
            }, context);

            return {
                name,
                span
            };
        });
    }, {
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, RemoteModulesDeferredRegistrationsUpdateCompletedEvent, (payload: unknown) => {
        if (deferredRegistrationsUpdateSpan) {
            deferredRegistrationsUpdateSpan.addEvent("remote-module-deferred-registrations-update-completed", {
                "app.squide.registration_count": (payload as RemoteModulesDeferredRegistrationsUpdateCompletedEventPayload).registrationCount
            });
        }

        if (remoteModuleDeferredRegistrationsUpdateSpan) {
            endActiveSpan(remoteModuleDeferredRegistrationsUpdateSpan);
        }
    }, {
        onError: handleUnmanagedError
    });

    // Can occur multiple times.
    addProtectedListener(runtime, RemoteModuleDeferredRegistrationUpdateFailedEvent, (payload: unknown) => {
        const registrationError = payload as RemoteModuleRegistrationError;

        if (remoteModuleDeferredRegistrationsUpdateSpan) {
            traceError(remoteModuleDeferredRegistrationsUpdateSpan.instance, registrationError);
        }
    }, {
        onError: handleUnmanagedError
    });
}

export function registerHoneycombInstrumentation(runtime: FireflyRuntime) {
    try {
        registerActiveSpanStack();

        // Dynamically registering this request hook function to nest the HTTP requests
        // of squide bootstrapping under the appropriate Honeycomb span.
        runtime.honeycombInstrumentationClient?.registerFetchRequestHook(createOverrideFetchRequestSpanWithActiveSpanContext(runtime.logger));

        registerTrackingListeners(runtime);

        runtime.logger.information("[squide] Honeycomb instrumentation is registered.");
    } catch (error: unknown) {
        runtime.logger
            .withText("[squide] An error occurred while registering Honeycomb instrumentation:")
            .withError(error as Error)
            .error();
    }
}
