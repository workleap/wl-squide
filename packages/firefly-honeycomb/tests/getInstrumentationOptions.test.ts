import type { FetchInstrumentationConfig } from "@opentelemetry/instrumentation-fetch";
import { FireflyRuntime } from "@squide/firefly";
import { describe, test, vi } from "vitest";
import { __clearOverrideFetchRequestSpanWithActiveSpanContextMock, __setOverrideFetchRequestSpanWithActiveSpanContextMock } from "../src/activeSpan.ts";
import { getInstrumentationOptions } from "../src/registerHoneycombInstrumentation.ts";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function removeInstrumentationVersionsForSnapshot(options: any) {
    if (Array.isArray(options.instrumentations)) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        options.instrumentations.forEach(x => {
            if (x["instrumentationVersion"]) {
                delete x["instrumentationVersion"];
            }

            if (x["version"]) {
                delete x["version"];
            }

            if (x["_logger"] && x["_logger"]["version"]) {
                delete x["_logger"]["version"];
            }

            if (x["_tracer"] && x["_tracer"]["version"]) {
                delete x["_tracer"]["version"];
            }
        });
    }

    return options;
}

test.concurrent("when debug is true", ({ expect }) => {
    const runtime = new FireflyRuntime();

    const result = getInstrumentationOptions(runtime, {
        debug: true,
        apiKey: "123"
    });

    const cleanedResult = removeInstrumentationVersionsForSnapshot({
        ...result,
        fetchInstrumentation: result.fetchInstrumentation ? result.fetchInstrumentation({}) : undefined
    });

    expect(cleanedResult).toMatchSnapshot();
});

test.concurrent("when debug is false", ({ expect }) => {
    const runtime = new FireflyRuntime();

    const result = getInstrumentationOptions(runtime, {
        debug: false,
        apiKey: "123"
    });

    const cleanedResult = removeInstrumentationVersionsForSnapshot({
        ...result,
        fetchInstrumentation: result.fetchInstrumentation ? result.fetchInstrumentation({}) : undefined
    });

    expect(cleanedResult).toMatchSnapshot();
});

test.concurrent("when the runtime mode is \"development\"", ({ expect }) => {
    const runtime = new FireflyRuntime({
        mode: "development"
    });

    const result = getInstrumentationOptions(runtime, {
        apiKey: "123"
    });

    const cleanedResult = removeInstrumentationVersionsForSnapshot({
        ...result,
        fetchInstrumentation: result.fetchInstrumentation ? result.fetchInstrumentation({}) : undefined
    });

    expect(cleanedResult).toMatchSnapshot();
});

test.concurrent("when the runtime mode is \"production\"", ({ expect }) => {
    const runtime = new FireflyRuntime({
        mode: "production"
    });

    const result = getInstrumentationOptions(runtime, {
        apiKey: "123"
    });

    const cleanedResult = removeInstrumentationVersionsForSnapshot({
        ...result,
        fetchInstrumentation: result.fetchInstrumentation ? result.fetchInstrumentation({}) : undefined
    });

    expect(cleanedResult).toMatchSnapshot();
});

describe("fetchInstrumentation", () => {
    test.concurrent("when fetchInstrumentation is false, return false", ({ expect }) => {
        const runtime = new FireflyRuntime();

        const result = getInstrumentationOptions(runtime, {
            fetchInstrumentation: false,
            apiKey: "123"
        });

        expect(result.fetchInstrumentation).toBe(false);
    });

    test.concurrent("when fetchInstrumentation is a function, call the function with the augmented options", ({ expect }) => {
        const runtime = new FireflyRuntime();

        const mock = vi.fn();

        const result = getInstrumentationOptions(runtime, {
            fetchInstrumentation: mock,
            apiKey: "123"
        });

        // Simulate calling the "registerHoneycombInstrumentation" function.
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        result.fetchInstrumentation({});

        expect(mock).toHaveBeenCalledTimes(1);
        expect(mock).toHaveBeenCalledWith(expect.objectContaining({
            requestHook: expect.any(Function)
        }));
    });

    test.concurrent("when fetchInstrumentation is not provided, requestHook is the active span override function", ({ expect }) => {
        const runtime = new FireflyRuntime();

        const result = getInstrumentationOptions(runtime, {
            apiKey: "123"
        });

        expect(result.fetchInstrumentation).toBeDefined();

        // Simulate calling the "registerHoneycombInstrumentation" function.
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const fetchOptions = result.fetchInstrumentation({}) as FetchInstrumentationConfig;

        expect(fetchOptions.requestHook).toBeDefined();
    });

    test.concurrent("when the base honeycomb instrumentation library configure a default requestHook, merge the base function with the active span override function", ({ expect }) => {
        const runtime = new FireflyRuntime();

        const baseConfigMock = vi.fn();
        const activeSpanMock = vi.fn();

        __setOverrideFetchRequestSpanWithActiveSpanContextMock(activeSpanMock);

        const result = getInstrumentationOptions(runtime, {
            apiKey: "123"
        });

        // Simulate calling the "registerHoneycombInstrumentation" function.
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const { requestHook } = result.fetchInstrumentation({
            requestHook: baseConfigMock
        });

        requestHook();

        expect(baseConfigMock).toHaveBeenCalledTimes(1);
        expect(activeSpanMock).toHaveBeenCalledTimes(1);

        __clearOverrideFetchRequestSpanWithActiveSpanContextMock();
    });
});
