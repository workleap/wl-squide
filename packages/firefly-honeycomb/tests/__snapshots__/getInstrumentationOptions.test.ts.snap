// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when debug is false 1`] = `
{
  "apiKey": "123",
  "debug": false,
  "fetchInstrumentation": {
    "requestHook": [Function],
  },
}
`;

exports[`when debug is true 1`] = `
{
  "apiKey": "123",
  "debug": true,
  "fetchInstrumentation": {
    "requestHook": [Function],
  },
}
`;

exports[`when the runtime mode is "development" 1`] = `
{
  "apiKey": "123",
  "debug": true,
  "fetchInstrumentation": {
    "requestHook": [Function],
  },
}
`;

exports[`when the runtime mode is "production" 1`] = `
{
  "apiKey": "123",
  "debug": false,
  "fetchInstrumentation": {
    "requestHook": [Function],
  },
}
`;
